src/miniscript/context.rs

```diff
@@ -0,0 +1,200 @@
+// Miniscript
+// Written in 2019 by
+//     Sanket Kanjalkar and Andrew Poelstra
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use std::fmt;
+use {Miniscript, MiniscriptKey, Terminal};
+
+/// Error for Script Context
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum ScriptContextError {
+    /// Script Context does not permit PkH for non-malleability
+    /// It is not possible to estimate the pubkey size at the creation
+    /// time because of uncompressed pubkeys
+    MalleablePkH,
+    /// Script Context does not permit OrI for non-malleability
+    /// Legacy fragments allow non-minimal IF which results in malleability
+    MalleableOrI,
+    /// Script Context does not permit DupIf for non-malleability
+    /// Legacy fragments allow non-minimal IF which results in malleability
+    MalleableDupIf,
+    /// Only Compressed keys allowed under current descriptor
+    /// Segwitv0 fragments do not allow uncompressed pubkeys
+    CompressedOnly,
+}
+
+impl fmt::Display for ScriptContextError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ScriptContextError::MalleablePkH => write!(f, "PkH is malleable under Legacy rules"),
+            ScriptContextError::MalleableOrI => write!(f, "OrI is malleable under Legacy rules"),
+            ScriptContextError::MalleableDupIf => {
+                write!(f, "DupIf is malleable under Legacy rules")
+            }
+            ScriptContextError::CompressedOnly => {
+                write!(f, "Uncompressed pubkeys not allowed in segwit context")
+            }
+        }
+    }
+}
+
+pub trait ScriptContext:
+    fmt::Debug + Clone + Ord + PartialOrd + Eq + PartialEq + private::Sealed
```

Comment:
Lol, do we actually need any of these extra traits beyond `private::Sealed`?

Reply:
Turns out we do. Rust is complaining that it can't compare, equate without those